packages/flutter/lib/src/material_animated_icons/animated_icons.dart

```diff
@@ -0,0 +1,269 @@
+  /// for setting the current theme's brightness.
+  final Color color;
+
+  /// The size of the icon in logical pixels.
+  ///
+  /// Icons occupy a square with width and height equal to size.
+  ///
+  /// Defaults to the current [IconTheme] size.
+  final double size;
+
+  /// The icon to display. Available icons are listed in [AnimatedIcons].
+  final AnimatedIconData icon;
+
+  /// Semantic label for the icon.
+  ///
+  /// This would be read out in accessibility modes (e.g TalkBack/VoiceOver).
+  /// This label does not show in the UI.
+  ///
+  /// See also:
+  ///
+  ///  * [Semantics.label], which is set to [semanticLabel] in the underlying
+  ///    [Semantics] widget.
+  final String semanticLabel;
+
+  /// The text direction to use for rendering the icon.
+  ///
+  /// If this is null, the ambient [Directionality] is used instead.
+  ///
+  /// If the text diection is [TextDirection.rtl], the icon will be mirrored
+  /// horizontally (e.g back arrow will point right).
+  final TextDirection textDirection;
+
+  static final _UiPathFactory _pathFactory = () => new ui.Path();
+
+  @override
+  Widget build(BuildContext context) {
+    final _AnimatedIconData iconData = icon;
+    final IconThemeData iconTheme = IconTheme.of(context);
+    final double iconSize = size ?? iconTheme.size;
+    final TextDirection textDirection = this.textDirection ?? Directionality.of(context);
+    return new Semantics(
+      label: semanticLabel,
+      child: new CustomPaint(
+        size: new Size(iconSize, iconSize),
+        painter: new _AnimatedIconPainter(
+          iconData.paths,
+          progress,
+          color ?? iconTheme.color,
+          iconSize / iconData.size.bottomRight(const Offset(0.0, 0.0)).dx,
+          textDirection == TextDirection.rtl,
+          _pathFactory,
+        ),
+      ),
+    );
+  }
+}
+
+typedef ui.Path _UiPathFactory();
+
+class _AnimatedIconPainter extends CustomPainter {
+  _AnimatedIconPainter(
+    this.paths,
+    this.progress,
+    this.color,
+    this.scale,
+    this.shouldMirror,
+    this.uiPathFactory,
+  ) : super(repaint: progress);
+
+  // This list is assumed to be immutable, changes to the contents of the list
+  // will not trigger a redraw as shouldRepaint will keep returning false.
+  final List<_PathFrames> paths;
+  final Animation<double> progress;
+  final Color color;
+  final double scale;
+  /// If this is true the image will be mirrored horizontally.
+  final bool shouldMirror;
+  final _UiPathFactory uiPathFactory;
+
+  @override
+  void paint(ui.Canvas canvas, Size size) {
+    canvas.scale(scale, scale);
+    if (shouldMirror) {
+      canvas.rotate(math.pi);
+      canvas.translate(-size.width, -size.height);
+    }
+
+    for (_PathFrames path in paths)
+      path.paint(canvas, color, uiPathFactory, progress.value.clamp(0.0, 1.0));
+  }
+
+
+  @override
+  bool shouldRepaint(_AnimatedIconPainter oldDelegate) {
+    return oldDelegate.progress.value != progress.value
```

Comment:
compare the progress, not the values

Reply:
following offline discussion in this case the value is the only thing out of progress that affects what is painted. So we keep the check as-is.

Reply:
This is still wrong, actually, though I can't quite figure out how to fix it.

Consider the case where the AnimatedIcon widget is being created in an AnimatedBuilder, so that it is rebuilt each time the ticker ticks. Each time, we'll check shouldRepaint, and the value will be the same in both delegates, since it's the same controller. So we won't repaint.

Reply:
(we figured out off-line that actually this is fine because in that case the ticker will trigger markNeedsPaint even though shouldRepaint is false)